### ISSUE TITLE
Inquiry on Integrating Homomorphic Encryption into the APPFL Framework

### ISSUE THREAD

**Issue by zhaozhao626** [APPFL/APPFL#170]
I recently came across your paper "APPFL: Open-Source Software Framework for Privacy-Preserving Federated Learning" and was impressed by your work. The APPFL framework provides a robust tool for implementing privacy-preserving federated learning, which is of significant importance for domains where data privacy is paramount, such as biomedicine and smart grids.

I am considering using the APPFL framework for my research project and am interested in integrating Homomorphic Encryption (HE) to further enhance the privacy protection of data. I understand that HE allows computations to be performed on encrypted data without decryption, which would be invaluable for safeguarding sensitive information from clients. However, I have a few questions regarding how to implement this technology within the existing APPFL framework, and I hope to receive your guidance.

Specifically, I have the following inquiries:

Which homomorphic encryption library would you recommend for use within the APPFL framework, and how can it be combined with existing privacy mechanisms such as Differential Privacy?
What components on the server and client sides would I need to modify to support Homomorphic Encryption?
Homomorphic Encryption can significantly increase computational overhead. Do you have any optimization suggestions or best practices to mitigate this impact?
After integrating Homomorphic Encryption, how should I evaluate the system's security and performance?
I look forward to your response and hope to successfully integrate Homomorphic Encryption into the APPFL framework with your assistance. If possible, I would also like to contribute to the APPFL community and participate in any related discussions.

Thank you for your time, and I eagerly await your reply.

**Comment by Zilinghan** [comment]
Hi, thanks for your question! 
1. To use homomorphic encryption, I recommend using the following repo https://github.com/OpenMined/TenSEAL, which is built on top of Microsoft SEAF for tensor. You can use it to encrypt any tensors before transmission from the client to the server, and decrypt the tensor back once receiving tensors from the server.
2. As HE already encrypts the data, I wonder if it is still necessary to do DP. From my point of view, if the model gradient can be accessible by hacker, then DP protects the model from inversion attack. For example, see section 2.4 of this [paper](https://arxiv.org/pdf/2312.08701.pdf).
3. As HE preservers operations such as addition and multiplication, I think the most important thing is to make sure the data format from the server and clients are well aligned during the aggregation. Also, you need to add encryption and decryption module at the client side.
4. I have no idea how to mitigate the computation overhead, as I think this is a tradeoff between security and efficiency.
5. Using Homomorphic Encryption, the security mainly depends on the HE algorithm itself and should be guaranteed. I think you could focus on the efficiency to measure the additional latency introduced by HE.

At the end, if you want easier communication, you are welcome to join our [slack workspace](https://join.slack.com/t/appflhq/shared_invite/zt-2ef419ydg-u2PqP8LzQLp7B2BFYFwy6g) and talk to APPFL Dev

**Comment by zhaozhao626** [comment]
I wanted to express my sincere gratitude for your prompt and insightful response regarding the integration of Homomorphic Encryption (HE) into the APPFL framework. 

I have reviewed the TenSEAL repository you recommended, which is built on Microsoft SEAL for tensors. I am considering using this library to encrypt tensors before they are transmitted from the client to the server, and then decrypt them upon reception. **However, I am unsure about the specific steps and functions to call within the APPFL framework to achieve this integration.**

Are there any specific functions or modules in APPFL that I should focus on to facilitate the integration of HE?

**Comment by Zilinghan** [comment]
I think a great first step to write two functions `homo_encrpytion` and `homo_decryption` with the function interfaces to encrypt the model state dictionary
`def homo_encrpytion(model: torch.nn.Module, **kwargs) -> Dict[str, Any]:`
`def homo_decrpytion(encrypted_model: Dict[str, Any], **kwargs) -> Dict[str, Any]:`

**Comment by zhaozhao626** [comment]
Your suggestion to create homo_encryption and homo_decryption functions is a great starting point, and **I am excited to delve into the implementation.**

If you have any additional insights or recommendations on best practices for implementing these functions, I would be more than happy to hear them. Your guidance has been incredibly helpful, and **I am eager to ensure that our approach aligns with the latest standards in the field of homomorphic encryption.**

**Thank you once again for your support.** I will keep you updated on our progress and would be glad to discuss any further ideas or questions that may arise during the development process.

**Comment by kibaekkim** [comment]
@zhaozhao626 I believe that a good starter should be following these steps:
1. install for developers: https://appfl.readthedocs.io/en/latest/install/index.html#developer-installation
2. make sure to read: https://appfl.readthedocs.io/en/latest/developer/index.html#developers
3. take one example and run it
4. point out from the example where you want to apply HE.

**Comment by kibaekkim** [comment]
> At the end, if you want easier communication, you are welcome to join our [Discord community](https://discord.gg/7ghjQbGcc4) and talk to APPFL Dev

@zhaozhao626 Consider to join the community for more interactive help.
